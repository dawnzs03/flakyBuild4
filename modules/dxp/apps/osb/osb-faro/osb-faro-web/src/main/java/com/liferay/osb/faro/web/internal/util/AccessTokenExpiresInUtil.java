/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

package com.liferay.osb.faro.web.internal.util;

/**
 * @author Ivica Cardic
 */
public class AccessTokenExpiresInUtil {

	public static long getExpiresIn() {
		return _expiresIn;
	}

	public static void removeExpiresIn() {
		_expiresIn = 0;
	}

	public static void setExpiresIn(long expiresIn) {
		_expiresIn = expiresIn;
	}

	private static long _expiresIn;

}