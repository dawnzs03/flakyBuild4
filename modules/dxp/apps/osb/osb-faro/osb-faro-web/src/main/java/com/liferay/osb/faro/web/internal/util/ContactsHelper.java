/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

package com.liferay.osb.faro.web.internal.util;

import com.liferay.osb.faro.engine.client.ContactsEngineClient;
import com.liferay.osb.faro.engine.client.constants.FieldMappingConstants;
import com.liferay.osb.faro.engine.client.model.Individual;
import com.liferay.osb.faro.engine.client.model.IndividualSegment;
import com.liferay.osb.faro.model.FaroProject;
import com.liferay.osb.faro.web.internal.constants.FaroConstants;
import com.liferay.osb.faro.web.internal.model.display.contacts.IndividualDisplay;
import com.liferay.osb.faro.web.internal.model.display.contacts.IndividualSegmentDisplay;
import com.liferay.osb.faro.web.internal.model.display.main.FaroEntityDisplay;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Shinn Lok
 */
@Component(service = ContactsHelper.class)
public class ContactsHelper {

	public FaroEntityDisplay getContactsEntityDisplay(
		FaroProject faroProject, String contactsEntityId, int type) {

		if (type == FaroConstants.TYPE_INDIVIDUAL) {
			Individual individual = _contactsEngineClient.getIndividual(
				faroProject, contactsEntityId, null);

			return new IndividualDisplay(individual);
		}

		IndividualSegment individualSegment =
			_contactsEngineClient.getIndividualSegment(
				faroProject, contactsEntityId, false);

		return new IndividualSegmentDisplay(individualSegment);
	}

	public String getOwnerType(int type) {
		if (type == FaroConstants.TYPE_ACCOUNT) {
			return FieldMappingConstants.OWNER_TYPE_ACCOUNT;
		}
		else if (type == FaroConstants.TYPE_INDIVIDUAL) {
			return FieldMappingConstants.OWNER_TYPE_INDIVIDUAL;
		}

		return FieldMappingConstants.OWNER_TYPE_INDIVIDUAL_SEGMENT;
	}

	@Reference
	private ContactsEngineClient _contactsEngineClient;

}