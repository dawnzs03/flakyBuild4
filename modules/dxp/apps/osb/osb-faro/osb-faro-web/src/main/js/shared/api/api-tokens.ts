import sendRequest from 'shared/util/request';
import {RESTParams} from 'shared/types';

export function generate({
	expiresIn,
	groupId
}: RESTParams & {expiresIn: string}) {
	return sendRequest({
		method: 'POST',
		path: `main/${groupId}/oauth2/tokens/new?expiresIn=${expiresIn}`
	});
}

export function search({groupId}: RESTParams) {
	return sendRequest({
		method: 'GET',
		path: `main/${groupId}/oauth2/tokens`
	});
}

export function revoke({groupId, token}: RESTParams & {token: string}) {
	return sendRequest({
		method: 'POST',
		path: `main/${groupId}/oauth2/tokens/${token}/revoke`
	});
}
