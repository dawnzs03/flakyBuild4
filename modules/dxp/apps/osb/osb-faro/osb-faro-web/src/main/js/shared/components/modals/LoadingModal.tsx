import ClayIcon from '@clayui/icon';
import getCN from 'classnames';
import Loading from 'shared/components/Loading';
import Modal from 'shared/components/modal';
import React from 'react';

interface ILoadingModalProps extends React.HTMLAttributes<HTMLDivElement> {
	icon?: string;
	message?: string;
	title?: string;
}

const LoadingModal: React.FC<ILoadingModalProps> = ({
	className,
	icon,
	message = Liferay.Language.get('loading'),
	title
}) => (
	<Modal className={getCN('loading-modal-root', className)} size='sm'>
		{title && <h1 className='title'>{title}</h1>}

		<div className='icon-container'>
			{icon ? (
				<ClayIcon className='icon-root icon-size-xl' symbol={icon} />
			) : (
				<Loading />
			)}
		</div>

		{message && <p className='message'>{message}</p>}
	</Modal>
);

export default LoadingModal;
