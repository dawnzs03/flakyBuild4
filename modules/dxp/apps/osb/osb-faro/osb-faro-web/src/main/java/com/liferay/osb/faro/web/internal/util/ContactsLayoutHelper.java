/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

package com.liferay.osb.faro.web.internal.util;

import com.liferay.osb.faro.contacts.model.ContactsCardTemplate;
import com.liferay.osb.faro.contacts.service.ContactsCardTemplateLocalService;
import com.liferay.osb.faro.web.internal.card.template.ContactsCardTemplateManagerHelper;
import com.liferay.osb.faro.web.internal.model.display.contacts.card.template.ContactsLayoutTemplateSettingDisplay;
import com.liferay.petra.string.StringBundler;
import com.liferay.petra.string.StringPool;
import com.liferay.portal.kernel.model.UserConstants;

import java.util.ArrayList;
import java.util.List;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Matthew Kong
 */
@Component(service = ContactsLayoutHelper.class)
public class ContactsLayoutHelper {

	public String addContactsCardTemplateIds(long groupId, int type)
		throws Exception {

		List<List<ContactsLayoutTemplateSettingDisplay>>
			contactsLayoutTemplateSettingDisplaysList = new ArrayList<>();

		for (int[] contactsCardTemplateTypes :
				_contactsCardTemplateManagerUtil.
					getDefaultContactsCardTemplateTypes(type)) {

			List<ContactsLayoutTemplateSettingDisplay>
				contactsLayoutTemplateSettingDisplays = new ArrayList<>();

			for (int contactsCardTemplateType : contactsCardTemplateTypes) {
				ContactsCardTemplate contactsCardTemplate =
					_contactsCardTemplateLocalService.addContactsCardTemplate(
						groupId, UserConstants.USER_ID_DEFAULT,
						_contactsCardTemplateManagerUtil.getDefaultName(
							contactsCardTemplateType),
						_contactsCardTemplateManagerUtil.getDefaultSettings(
							contactsCardTemplateType),
						contactsCardTemplateType);

				contactsLayoutTemplateSettingDisplays.add(
					new ContactsLayoutTemplateSettingDisplay(
						contactsCardTemplate.getContactsCardTemplateId(), 0));
			}

			contactsLayoutTemplateSettingDisplaysList.add(
				contactsLayoutTemplateSettingDisplays);
		}

		return JSONUtil.writeValueAsString(
			contactsLayoutTemplateSettingDisplaysList);
	}

	public String addHeaderContactsCardTemplateIds(long groupId, int type) {
		int[] contactsCardTemplateTypes =
			_contactsCardTemplateManagerUtil.
				getDefaultHeaderContactsCardTemplateTypes(type);

		if (contactsCardTemplateTypes.length == 0) {
			return StringPool.BLANK;
		}

		StringBundler sb = new StringBundler(
			contactsCardTemplateTypes.length * 2);

		for (int contactsCardTemplateType : contactsCardTemplateTypes) {
			ContactsCardTemplate contactsCardTemplate =
				_contactsCardTemplateLocalService.addContactsCardTemplate(
					groupId, UserConstants.USER_ID_DEFAULT,
					_contactsCardTemplateManagerUtil.getDefaultName(
						contactsCardTemplateType),
					_contactsCardTemplateManagerUtil.getDefaultSettings(
						contactsCardTemplateType),
					contactsCardTemplateType);

			sb.append(contactsCardTemplate.getContactsCardTemplateId());

			sb.append(StringPool.COMMA);
		}

		sb.setIndex(sb.index() - 1);

		return sb.toString();
	}

	@Reference
	private ContactsCardTemplateLocalService _contactsCardTemplateLocalService;

	@Reference
	private ContactsCardTemplateManagerHelper _contactsCardTemplateManagerUtil;

}