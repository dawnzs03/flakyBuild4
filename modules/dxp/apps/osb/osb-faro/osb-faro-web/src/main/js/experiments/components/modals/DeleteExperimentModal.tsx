import BaseModal from 'experiments/components/modals/BaseModal';
import React from 'react';
import {addAlert} from 'shared/actions/alerts';
import {Alert} from 'shared/types';
import {connect} from 'react-redux';
import {EXPERIMENT_DELETE_MUTATION} from 'experiments/queries/ExperimentMutation';
import {Routes, toRoute} from 'shared/util/router';
import {useHistory, useParams} from 'react-router-dom';
import {useMutation} from '@apollo/react-hooks';

const DeleteExperimentModal = ({addAlert, experimentId, observer, onClose}) => {
	const {channelId, groupId} = useParams();
	const history = useHistory();
	const [mutate] = useMutation(EXPERIMENT_DELETE_MUTATION);

	const onSubmit = () =>
		mutate({
			variables: {
				experimentId
			}
		});

	const onSuccess = () => {
		addAlert({
			alertType: Alert.Types.Success,
			message: Liferay.Language.get('the-test-has-been-deleted')
		});

		history.push(toRoute(Routes.TESTS, {channelId, groupId}));
	};

	return (
		<BaseModal
			observer={observer}
			onClose={onClose}
			onSubmit={onSubmit}
			onSuccess={onSuccess}
			status='warning'
			submitMessage={Liferay.Language.get('delete')}
			title={Liferay.Language.get('deleting-test')}
		>
			<p className='font-weight-bold text-secondary'>
				{Liferay.Language.get(
					'are-you-sure-you-want-to-delete-this-test'
				)}
			</p>

			<p className='text-secondary'>
				{Liferay.Language.get(
					'you-will-permanently-lose-all-test-data-and-results-collected-from-this-test'
				)}{' '}
				{Liferay.Language.get('you-will-not-be-able-to-run-this-test')}{' '}
				{Liferay.Language.get(
					'you-will-not-be-able-to-undo-this-operation'
				)}
			</p>
		</BaseModal>
	);
};

export default connect(null, {addAlert})(DeleteExperimentModal);
