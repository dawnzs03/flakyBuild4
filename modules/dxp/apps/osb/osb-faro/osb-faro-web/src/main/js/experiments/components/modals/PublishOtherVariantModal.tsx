import BaseModal from 'experiments/components/modals/BaseModal';
import DXPLinkButton from './DXPLinkButton';
import PublishVariant from './PublishVariant';
import React, {useState} from 'react';
import {EXPERIMENT_MUTATION} from 'experiments/queries/ExperimentMutation';
import {getVariantLink, makeAllRefetch} from 'experiments/util/experiments';
import {Option, Picker} from '@clayui/core';
import {useMutation} from '@apollo/react-hooks';
import {useStateValue} from 'experiments/state';

const ChooseVariant = ({dxpVariant, dxpVariants, link, setDxpVariant}) => (
	<>
		<div className='mb-3 font-size-md font-weight-semibold'>
			{Liferay.Language.get('which-variant-do-you-want-to-publish')}
		</div>

		<Picker
			className='mb-3'
			onSelectionChange={value =>
				setDxpVariant(
					dxpVariants.find(({dxpVariantId}) => dxpVariantId === value)
				)
			}
			selectedKey={dxpVariant ? dxpVariant.dxpVariantId : ''}
		>
			<Option key='select-a-variant'>
				{Liferay.Language.get('select-a-variant')}
			</Option>
			{dxpVariants.map(({dxpVariantId, dxpVariantName}) => (
				<Option key={dxpVariantId}>{dxpVariantName}</Option>
			))}
		</Picker>

		<DXPLinkButton disabled={!dxpVariant} href={link} />

		<div className='mt-3'>
			{Liferay.Language.get(
				'this-will-publish-the-variant-as-the-live-experience-for-the-target-segment-that-was-chosen-for-this-test'
			)}
		</div>
	</>
);

const PublishOtherVariantModel = ({
	dxpVariants,
	experimentId,
	observer,
	onClose,
	pageURL,
	title
}) => {
	const [{allRefetch}]: any = useStateValue();
	const [dxpVariant, setDxpVariant] = useState(null);
	const [currentStep, setStep] = useState(1);
	const [mutate] = useMutation(EXPERIMENT_MUTATION);

	const onSubmit =
		currentStep === 1
			? () => setStep(2)
			: () =>
					mutate({
						variables: {
							experimentId,
							publishedDXPVariantId: dxpVariant.dxpVariantId,
							status: 'COMPLETED'
						}
					});
	const closeFn = currentStep === 1 ? () => null : onClose;
	const onCancel = currentStep === 1 ? onClose : () => setStep(1);

	const labelSubmitButton =
		currentStep === 1
			? Liferay.Language.get('next')
			: Liferay.Language.get('publish-as-experience');

	const labelCancelButton =
		currentStep === 1
			? Liferay.Language.get('cancel')
			: Liferay.Language.get('back');

	const onSuccess =
		currentStep === 1 ? () => {} : () => makeAllRefetch(allRefetch);

	const link = dxpVariant
		? getVariantLink({id: dxpVariant.dxpVariantId, pageURL})
		: '';

	return (
		<BaseModal
			cancelMessage={labelCancelButton}
			disabled={!dxpVariant}
			observer={observer}
			onCancel={onCancel}
			onClose={closeFn}
			onSubmit={onSubmit}
			onSuccess={onSuccess}
			submitMessage={labelSubmitButton}
			title={title}
		>
			{currentStep === 1 ? (
				<ChooseVariant
					dxpVariant={dxpVariant}
					dxpVariants={dxpVariants}
					link={link}
					setDxpVariant={setDxpVariant}
				/>
			) : (
				<PublishVariant
					dxpVariantName={dxpVariant.dxpVariantName}
					link={link}
				/>
			)}
		</BaseModal>
	);
};

export default PublishOtherVariantModel;
