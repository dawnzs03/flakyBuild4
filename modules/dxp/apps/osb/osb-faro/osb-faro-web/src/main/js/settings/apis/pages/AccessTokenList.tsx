import * as API from 'shared/api';
import Alerts, {AlertTypes} from 'shared/components/Alert';
import BasePage from 'settings/components/BasePage';
import Card from 'shared/components/Card';
import ClayButton from '@clayui/button';
import CopyButton from 'shared/components/CopyButton';
import GenerateTokenCard from '../components/GenerateTokenCard';
import Loading, {Align} from 'shared/components/Loading';
import moment from 'moment';
import React, {useState} from 'react';
import Table from 'shared/components/table';
import TokenCell from '../components/TokenCell';
import URLConstants from 'shared/util/url-constants';
import {AccessToken} from '../types';
import {addAlert} from 'shared/actions/alerts';
import {Alert} from 'shared/types';
import {ApisPath} from 'shared/util/url-constants';
import {close, modalTypes, open} from 'shared/actions/modals';
import {compose} from 'redux';
import {connect, ConnectedProps} from 'react-redux';
import {CUSTOM_DATE_FORMAT} from 'shared/util/date';
import {ENABLE_LAST_ACCESS_DATE, ExpirationPeriod} from 'shared/util/constants';
import {formatDateToTimeZone, getDateNow} from 'shared/util/date';
import {RootState} from 'shared/store';
import {sub} from 'shared/util/lang';
import {
	withAdminPermission,
	withError,
	withLoading,
	withQuery
} from 'shared/hoc';
import type {Column} from 'shared/components/table/Row';

export const isExpired = (expirationDate: string) =>
	moment.utc(expirationDate).isSameOrBefore(getDateNow());

const getTimestamp = (date: Date) =>
	Math.floor(new Date(date).getTime() / 1000);

const isIndefinite = ({createDate, expirationDate}) =>
	getTimestamp(expirationDate) - getTimestamp(createDate) ===
	Number(ExpirationPeriod.Indefinite);

const connector = connect(
	(store: RootState, {groupId}: {groupId: string}) => ({
		timeZoneId: store.getIn([
			'projects',
			groupId,
			'data',
			'timeZone',
			'timeZoneId'
		])
	}),
	{addAlert, close, open}
);

type PropsFromRedux = ConnectedProps<typeof connector>;

const TokenList: React.FC<
	{
		groupId: string;
		refetch: () => Promise<any>;
		tokens: AccessToken[];
	} & PropsFromRedux
> = ({addAlert, close, groupId, open, refetch, timeZoneId, tokens}) => {
	const [loading, setLoading] = useState(false);
	const [onCloseAlert, setOnCloseAlert] = useState(false);

	const tokenExpired = !!tokens.length && isExpired(tokens[0].expirationDate);

	const handleError = () => {
		setLoading(false);

		addAlert({
			alertType: Alert.Types.Error,
			message: Liferay.Language.get('error'),
			timeout: false
		});
	};

	const handleSuccess = message => {
		setLoading(false);

		addAlert({
			alertType: Alert.Types.Success,
			message
		});

		refetch();
	};

	return (
		<div className='col-xl-8 pl-0'>
			{tokenExpired && !onCloseAlert && (
				<Alerts
					iconSymbol='warning-full'
					onClose={() => setOnCloseAlert(true)}
					title={Liferay.Language.get('warning')}
					type={AlertTypes.Warning}
				>
					{Liferay.Language.get('expired-token-warning')}
				</Alerts>
			)}

			{(tokenExpired || !tokens.length) && (
				<GenerateTokenCard
					groupId={groupId}
					onError={handleError}
					onSuccess={handleSuccess}
					token={tokens[0]?.token}
				/>
			)}

			<Card>
				<Card.Body>
					<div className='align-items-start d-flex flex-column justify-content-between'>
						<h4 className='mb-4'>
							{Liferay.Language.get('token-information')}
						</h4>

						<h5>{Liferay.Language.get('root-endpoint')}</h5>

						<span className='text-secondary'>
							{window.location.origin + ApisPath}
						</span>
					</div>
				</Card.Body>

				{!!tokens.length && (
					<Table
						className='mb-0'
						columns={
							[
								{
									accessor: 'token',
									cellRenderer: TokenCell,
									label: Liferay.Language.get('token'),
									sortable: false
								},
								ENABLE_LAST_ACCESS_DATE && {
									accessor: 'lastAccessDate',
									dataFormatter: (val: string) =>
										formatDateToTimeZone(
											val,
											CUSTOM_DATE_FORMAT,
											timeZoneId
										),
									label: Liferay.Language.get('last-seen'),
									sortable: false
								},
								{
									accessor: 'createDate',
									dataFormatter: (val: string) =>
										formatDateToTimeZone(
											val,
											CUSTOM_DATE_FORMAT,
											timeZoneId
										),
									label: Liferay.Language.get('date-created'),
									sortable: false
								},
								{
									accessor: 'expirationDate',
									cellRenderer: ({data}) => {
										if (isIndefinite(data)) {
											return (
												<td>
													{Liferay.Language.get(
														'indefinite'
													)}
												</td>
											);
										}

										return (
											<td>
												{formatDateToTimeZone(
													data.expirationDate,
													CUSTOM_DATE_FORMAT,
													timeZoneId
												)}
											</td>
										);
									},
									label: Liferay.Language.get('expiration'),
									sortable: false
								}
							].filter(Boolean) as Column[]
						}
						items={tokens}
						renderInlineRowActions={({data: {token}}) => {
							if (tokenExpired) return null;

							return (
								<>
									<CopyButton
										displayType='secondary'
										text={token}
									/>

									<ClayButton
										className='button-root'
										disabled={loading}
										displayType='secondary'
										onClick={() => {
											open(
												modalTypes.CONFIRMATION_MODAL,
												{
													message: (
														<div className='text-secondary'>
															<div>
																<strong>
																	{Liferay.Language.get(
																		'are-you-sure-you-want-to-revoke-this-token'
																	)}
																</strong>
															</div>

															{Liferay.Language.get(
																'you-will-need-to-generate-a-new-token-to-continue-using-this-api'
															)}
														</div>
													),
													modalVariant:
														'modal-warning',
													onClose: close,
													onSubmit: () => {
														setLoading(true);

														API.apiTokens
															.revoke({
																groupId,
																token
															})
															.then(() =>
																handleSuccess(
																	Liferay.Language.get(
																		'token-successfully-revoked'
																	)
																)
															)
															.catch(handleError);
													},
													submitButtonDisplay:
														'warning',
													title: Liferay.Language.get(
														'revoke-token'
													),
													titleIcon: 'warning-full'
												}
											);
										}}
									>
										{Liferay.Language.get('revoke')}

										{loading && (
											<Loading align={Align.Right} />
										)}
									</ClayButton>
								</>
							);
						}}
						rowIdentifier='token'
					/>
				)}
			</Card>
		</div>
	);
};

const ListWithData = compose<any>(
	connector,
	withQuery(
		API.apiTokens.search,
		({groupId}: {groupId: string}) => ({groupId}),
		({data, ...otherParams}) => ({
			tokens: data,
			...otherParams
		})
	),
	withLoading(),
	withError({page: false})
)(TokenList);

interface IAccessTokenListProps {
	groupId: string;
}

export const AccessTokenList: React.FC<IAccessTokenListProps> = ({groupId}) => (
	<BasePage
		className='access-token-list-root'
		groupId={groupId}
		pageDescription={sub(
			Liferay.Language.get(
				'access-this-workspaces-data-via-api-using-an-access-token.-a-full-list-of-endpoints-is-available-in-the-x'
			),
			[
				<a
					href={URLConstants.APIOverviewDocumentationLink}
					key='API_OVERVIEW_DOCUMENTATION'
					target='_blank'
				>
					{Liferay.Language.get('documentation-fragment')}
				</a>
			],
			false
		)}
		pageTitle={Liferay.Language.get('access-tokens')}
	>
		<ListWithData groupId={groupId} />
	</BasePage>
);

export default withAdminPermission(AccessTokenList);
