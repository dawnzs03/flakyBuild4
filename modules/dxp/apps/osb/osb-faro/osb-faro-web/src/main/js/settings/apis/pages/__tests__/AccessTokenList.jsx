import 'test/mock-modal';
import * as API from 'shared/api';
import * as data from 'test/data';
import mockStore from 'test/mock-store';
import Promise from 'metal-promise';
import React from 'react';
import {AccessTokenList} from '../AccessTokenList';
import {cleanup, fireEvent, getByText, render} from '@testing-library/react';
import {getISODate} from 'shared/util/date';
import {mockGetDateNow} from 'test/mock-date';
import {open} from 'shared/actions/modals';
import {Provider} from 'react-redux';
import {StaticRouter} from 'react-router';

jest.unmock('react-dom');

const defaultProps = {
	groupId: '23'
};

const DefaultComponent = props => (
	<StaticRouter>
		<Provider store={mockStore()}>
			<AccessTokenList {...defaultProps} {...props} />
		</Provider>
	</StaticRouter>
);

describe('AccessTokenList', () => {
	beforeAll(() => {
		mockGetDateNow(data.getTimestamp(0));
	});

	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(container).toMatchSnapshot();
	});

	it('should display the card with the options to create a new token if there is no token', () => {
		API.apiTokens.search.mockReturnValueOnce(Promise.resolve([]));

		const {container, queryByTestId} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(queryByTestId('generate-token-button')).toBeTruthy();
		expect(container.querySelector('.table-root')).toBeNull();
	});

	it('should show the generated token in a list and the "Generate Token" button should no longer be visible', () => {
		API.apiTokens.search.mockReturnValueOnce(Promise.resolve([]));

		const {container, getByTestId, queryByTestId} = render(
			<DefaultComponent />
		);

		jest.runAllTimers();

		expect(container.querySelector('.table-root')).toBeNull();
		expect(queryByTestId('generate-token-button')).toBeTruthy();

		fireEvent.click(getByTestId('generate-token-button'));

		jest.runAllTimers();

		expect(container.querySelector('.table-root')).toMatchSnapshot();
		expect(queryByTestId('generate-token-button')).toBeNull();
	});

	it('should open a modal to confirm revoking a token', () => {
		const {getByText} = render(<DefaultComponent />);

		jest.runAllTimers();

		fireEvent.click(getByText('Revoke'));

		jest.runAllTimers();

		expect(open).toBeCalled();
	});
	it('should display the "Generate Token" card  above the table if the token is expired', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					expirationDate: getISODate(data.getTimestamp(-1))
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(container.querySelector('.card-body'), 'Generate Token')
		).toBeTruthy();
		expect(container.querySelector('.table-root')).toMatchSnapshot();
	});

	it('should render the correct date on expiration date column when generated token is 30 days', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					createDate: '2023-05-11T19:35:28.338Z',
					expirationDate: '2023-06-12T19:35:28.000Z'
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(
				container.querySelector('td:nth-child(3)'),
				'Jun 12, 2023'
			)
		).toMatchSnapshot();
	});

	it('should render the correct date on expiration date column when generated token is 6 months', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					createDate: '2023-05-11T19:35:28.338Z',
					expirationDate: '2023-11-12T19:35:28.000Z'
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(
				container.querySelector('td:nth-child(3)'),
				'Nov 12, 2023'
			)
		).toMatchSnapshot();
	});

	it('should render the correct date on expiration date column when generated token is 1 year', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					createDate: '2023-05-11T19:35:28.338Z',
					expirationDate: '2024-05-12T19:35:28.000Z'
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(
				container.querySelector('td:nth-child(3)'),
				'May 12, 2024'
			)
		).toMatchSnapshot();
	});

	it('should render indefinite on expiration date column when generated token is indefinite', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					createDate: '2023-05-11T19:35:28.338Z',
					expirationDate: '2123-05-12T19:35:28.000Z'
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(container.querySelector('td:nth-child(3)'), 'Indefinite')
		).toMatchSnapshot();
	});
});
