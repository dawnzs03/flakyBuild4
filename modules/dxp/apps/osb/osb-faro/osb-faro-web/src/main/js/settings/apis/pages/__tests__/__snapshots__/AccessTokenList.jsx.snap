// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessTokenList should display the "Generate Token" card  above the table if the token is expired 1`] = `
<div
  class="table-responsive table-root flex-grow-1 mb-0"
>
  <table
    class="table table-autofit table-hover table-heading-nowrap table-nowrap"
  >
    <thead>
      <tr>
        <th
          class="table-head-title"
        >
          Token
        </th>
        <th
          class="table-head-title"
        >
          Date Created
        </th>
        <th
          class="table-head-title"
        >
          Expiration
        </th>
        <th />
      </tr>
    </thead>
    <tbody
      class="mb-0"
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          <span
            class="text-secondary mr-1"
          >
            Token ending in
          </span>
          <strong
            class="font-weight-bold"
          >
            4ABC
          </strong>
          <div
            class="label label-root ml-2 label-danger label-lg label-uppercase"
          >
            <span
              class="label-item"
            >
              Expired
            </span>
          </div>
        </td>
        <td>
          Jul 10, 2018
        </td>
        <td>
          Jul 09, 2018
        </td>
        <td
          class="row-inline-actions"
        >
          <span />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`AccessTokenList should render 1`] = `
<div>
  <div
    class="settings-base-page-root access-token-list-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Access Tokens
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Access this workspace's data via API using an access token. A full list of endpoints is available in the 
          <a
            href="https://learn.liferay.com/analytics-cloud/latest/en/apis.html"
            target="_blank"
          >
            documentation
          </a>
          .
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="col-xl-8 pl-0"
      >
        <div
          class="card card-root"
        >
          <div
            class="card-body"
          >
            <div
              class="align-items-start d-flex flex-column justify-content-between"
            >
              <h4
                class="mb-4"
              >
                Token Information
              </h4>
              <h5>
                Root Endpoint:
              </h5>
              <span
                class="text-secondary"
              >
                http://liferay.com/api/reports
              </span>
            </div>
          </div>
          <div
            class="table-responsive table-root flex-grow-1 mb-0"
          >
            <table
              class="table table-autofit table-hover table-heading-nowrap table-nowrap"
            >
              <thead>
                <tr>
                  <th
                    class="table-head-title"
                  >
                    Token
                  </th>
                  <th
                    class="table-head-title"
                  >
                    Date Created
                  </th>
                  <th
                    class="table-head-title"
                  >
                    Expiration
                  </th>
                  <th />
                </tr>
              </thead>
              <tbody
                class="mb-0"
              >
                <tr
                  class=""
                >
                  <td
                    class=""
                  >
                    <span
                      class="text-secondary mr-1"
                    >
                      Token ending in
                    </span>
                    <strong
                      class="font-weight-bold"
                    >
                      4ABC
                    </strong>
                  </td>
                  <td>
                    Jul 10, 2018
                  </td>
                  <td>
                    Aug 09, 2018
                  </td>
                  <td
                    class="row-inline-actions"
                  >
                    <span>
                      <button
                        aria-label="Click to Copy"
                        class="button-root btn btn-secondary"
                        data-clipboard-text="1234ABC"
                        data-tooltip-response="Copied"
                        title="Click to Copy"
                        type="button"
                      >
                        <svg
                          class="lexicon-icon lexicon-icon-copy icon-root"
                          role="presentation"
                        >
                          <use
                            xlink:href="#copy"
                          />
                        </svg>
                      </button>
                      <button
                        class="button-root btn btn-secondary"
                        type="button"
                      >
                        Revoke
                      </button>
                    </span>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AccessTokenList should render indefinite on expiration date column when generated token is indefinite 1`] = `
<td>
  Indefinite
</td>
`;

exports[`AccessTokenList should render the correct date on expiration date column when generated token is 1 year 1`] = `
<td>
  May 12, 2024
</td>
`;

exports[`AccessTokenList should render the correct date on expiration date column when generated token is 6 months 1`] = `
<td>
  Nov 12, 2023
</td>
`;

exports[`AccessTokenList should render the correct date on expiration date column when generated token is 30 days 1`] = `
<td>
  Jun 12, 2023
</td>
`;

exports[`AccessTokenList should show the generated token in a list and the "Generate Token" button should no longer be visible 1`] = `
<div
  class="table-responsive table-root flex-grow-1 mb-0"
>
  <table
    class="table table-autofit table-hover table-heading-nowrap table-nowrap"
  >
    <thead>
      <tr>
        <th
          class="table-head-title"
        >
          Token
        </th>
        <th
          class="table-head-title"
        >
          Date Created
        </th>
        <th
          class="table-head-title"
        >
          Expiration
        </th>
        <th />
      </tr>
    </thead>
    <tbody
      class="mb-0"
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          <span
            class="text-secondary mr-1"
          >
            Token ending in
          </span>
          <strong
            class="font-weight-bold"
          >
            4ABC
          </strong>
        </td>
        <td>
          Jul 10, 2018
        </td>
        <td>
          Aug 09, 2018
        </td>
        <td
          class="row-inline-actions"
        >
          <span>
            <button
              aria-label="Click to Copy"
              class="button-root btn btn-secondary"
              data-clipboard-text="1234ABC"
              data-tooltip-response="Copied"
              title="Click to Copy"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-copy icon-root"
                role="presentation"
              >
                <use
                  xlink:href="#copy"
                />
              </svg>
            </button>
            <button
              class="button-root btn btn-secondary"
              type="button"
            >
              Revoke
            </button>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
